Mainnet update

mds1 opened the pull request:
- Updates the sample scripts with mainnet info
- Updates the trigger interface to simplify trigger creation by already having `checkAndToggleTrigger` implemented
- Adds abstract interest rate model to inherit from to simplify process to create your own interest rate model
- Moves abstract contracts/interfaces into an `interfaces` folder
- Bumps solidity version to 0.8.5

mds1 committed:
Replace rinkeby setup with mainnet

mds1 committed:
Replace DAI with USDC as protection market default

mds1 committed:
Update package versions

mds1 committed:
Update eslint config

mds1 committed:
Cleanup utils and README

mds1 committed:
Update guides for forked mainnet

mds1 committed:
Update trigger contracts and guide

mds1 committed:
Update scripts to use ETH as collateral for borrowing

mds1 committed:
Update README, stop impersonating account after funding

mds1 committed:
Remove unsupported networks from config, bump solc version

mds1 committed:
Add abstract interest rate model contract, add interfaces folder

apbendi commented on guides/create-protection-market.ts:
```diff
@@ -23,45 +23,52 @@ const symbol = 'MOCK'; // trigger symbol
 
   // Deploy the trigger contract (last constructor parameter is specific to the trigger contract)
-  const trigger: Contract = await MockTriggerFactory.deploy(name, symbol, description, platformIds, recipient, false);
+  const trigger: Contract = await MockTriggerFactory.deploy(...triggerParams);
```
I might suggest moving the definition of `triggerParams` down to the line above this one, instead of on line 27. As it stands now, I have to go looking for their definition to figure out what's going on, especially with regards to the parenthetical in the comment on line 45.

mds1 replied:
Done in 4e1e403e8cba8022cf7f3be7fb7e7e4f9144dee5

apbendi commented on utils/utils.ts:
```diff
@@ -1,90 +1,118 @@
+  await hre.network.provider.request({ method: 'hardhat_impersonateAccount', params: [exchange] });
+  const signer = await hre.ethers.provider.getSigner(exchange);
+
+  if (tokenAddress === mainnetAddresses.ETH) {
```
I'd suggest adding a balance check for the amount of ETH/token requested to send vs. that which is held in this account before attempting the transfer, for the purpose of giving a helpful error message if the amount is insufficient. Without this, the user might be faced with a cryptic revert message depending on the token contract in question.

I'd suggest something along the lines of:

```
The mainnet address from which funds are confiscated for testing purposes does not have sufficient balance of ${currency} to fund the account.

Account: ${account}
Balance: ${balance}
Requested: ${requested}

Please request an amount the address is able to fund, or update the address used in utils.ts to a mainnet address that currently holds at least ${requested} ${currency}.
```

mds1 replied:
Done in 4e1e403e8cba8022cf7f3be7fb7e7e4f9144dee5

apbendi commented on utils/utils.ts:
```diff
@@ -1,90 +1,118 @@
+  const tokenAbi = ['function transfer(address,uint256) returns (bool)', 'function decimals() view returns (uint8)'];
+
+  // We impersonate the binance exchange account on mainnet as a source of tokens
+  const exchange = '0x28C6c06298d514Db089934071355E5743bf21d60';
```
Might be good to move this to the top of the file and comment it heavily. I'd also suggest renaming it from `exchange` to something like `mainnetFundingAccount` or something more descriptive. The fact this is an exchange account is an implementation detail here. If Binance ever deprecates this account, or if a user is ever testing a token they happen not to hold in it, they'll have to update it, so I think it's best to be as explicit as possible.

mds1 replied:
Renamed to `funderAddress` in 4e1e403e8cba8022cf7f3be7fb7e7e4f9144dee5

apbendi reviewed (commented):
Overall looks good! I left some suggestions. However, I'm unable to run any of the guides, as I'm getting this error:

```
TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at Object.exports.getContractAddress (cozy-developer-guides/utils/utils.ts:43:15)
    at main (cozy-developer-guides/guides/create-protection-market.ts:39:22)
```

Double checked my `.env` and that looks ok. Any ideas?

mds1 committed:
Add balance checks for funderAddress, move triggerParams definition

mds1 committed:
Remove outdated rinkeby addresses

mds1 commented:
TODO before merging: update mainnet addresses after deploy updates

mds1 committed:
Clean up scripts

- Remove stray console.log
- Fix balanceOf function signature
- Add return after a logFailure

mds1 committed:
Update mainnet contract addresses
